### ISSUE TITLE
Multiple BackupStorageLocations for one provider/plugin are flapping

### ISSUE THREAD

**Issue by Lirt** [vmware-tanzu/velero#5582]
Hello Velero team,

I'm a maintainer of [Velero community plugin for Openstack](https://github.com/Lirt/velero-plugin-for-openstack).

I'm trying to implement a new scenario where I want to have credential file with multiple clouds (let's say `cloud1` and `cloud2)` which are 2 completely different Openstack clouds (with different authentication and service client endpoints). I want to have 2 BackupStorageLocations where both of them will have different cloud in config - here is example:
```yaml
apiVersion: velero.io/v1
kind: BackupStorageLocation
metadata:
  name: bsl-cloud1
spec:
  accessMode: ReadWrite
  config:
    cloud: cloud1
  default: false
  objectStorage:
    bucket: bucket-in-cloud1
  provider: community.openstack.org/openstack
---
apiVersion: velero.io/v1
kind: BackupStorageLocation
metadata:
  name: bsl-cloud2
spec:
  accessMode: ReadWrite
  config:
    cloud: cloud2
  default: false
  objectStorage:
    bucket: bucket-in-cloud2
  provider: community.openstack.org/openstack
```

Now I should be able to create backups in both locations using 1 `Velero` instance with commands like this:
```bash
$ velero backup create --storage-location bsl-cloud1 backup-cloud1-01
$ velero backup create --storage-location bsl-cloud2 backup-cloud2-01
```

What happens is however that `Velero` is flapping between both clouds, trying to authenticate against first, list existing backups then the same for second cloud in a loop resulting in situation when sometimes I see only backups from `cloud1` and sometimes only from `cloud2`. Authentication and backup operations are successful.

If I understand correctly the BSL operations should be done sequentially, not in parallel, so using 1 client for provider and service should be OK. But reality is different and there is most likely a race condition.

I tried to rewrite the code to use 1 provider/service client per cloud (service/provider for authentication or object storage operations) and use proper one for each operation, but the plugin interface doesn't give me information about which cloud comes the request from and so It's not possible to identify for me which one to use. Thinking in go code:
```go
// I rewrote this
// type ObjectStore struct {
// 	client   *gophercloud.ServiceClient
// 	provider *gophercloud.ProviderClient
// 	log      logrus.FieldLogger
// }

// To this
type ObjectStore struct {
	client   map[string]*gophercloud.ServiceClient
	provider map[string]*gophercloud.ProviderClient
	log      logrus.FieldLogger
}
...
if cloud, ok := config["cloud"]; ok {
	log.Infof("Authentication will be done for cloud %v", cloud)
	err := utils.Authenticate(o.provider[cloud], "swift", config, b.log)
}

...
// Then in GetObject I don't have information from which BSL request comes or what is config of that BSL
func (o *ObjectStore) GetObject(bucket, key string) (io.ReadCloser, error) {
	object := objects.Download(o.client[???], bucket, key, nil)
        ...
}
```

Is there any way how to identify for which BSL is the operation meant for? Do you have any recommendations how to make this use-case scenario possible? 

The possible solution to identify operation is to change the plugin interface to include config or BSL name in input arguments. I understand this will possibly be a breaking change, but this is only one of the possible solutions I can think of.

```go
// This is official plugin interface
// config map[string]string is not passed to this of other functions and therefore 
// I'm not able to identify for which cloud the operation is supposed to be done 
// and use proper service client
GetObject(bucket, key string) (io.ReadCloser, error)

// If the config or name of BSL that operation is for was passed to all functions 
// from the interface it could be possible (https://github.com/vmware-tanzu/velero/blob/main/pkg/plugin/velero/object_store.go)
GetObject(bucket, key string, config map[string]string) (io.ReadCloser, error)
GetObject(bucket, key string, bsl string) (io.ReadCloser, error)
```

**Environment:**

- Velero version (use `velero version`): 1.9.2
- Kubernetes version (use `kubectl version`): 1.24.4
- Cloud provider or hardware configuration: Openstack

**Comment by blackpiglet** [comment]
I think the main problem you met is creating two backups without waiting them completed, and the two backups are handled by Velero at the same time. That causes some conflict, and make the two backups doesn't work as expected.

I think this should be a common problem for all cloud provider use cases, and this is not the expected behavior. I will take a look and try to reproduce it.

**Comment by Lirt** [comment]
Hi @blackpiglet, 

> I think the main problem you met is creating two backups without waiting them completed, and the two backups are handled by Velero at the same time. That causes some conflict, and make the two backups doesn't work as expected.

thank you for quick response. That's good to know - I will try to avoid this using different schedule times (also note it in my documentation).

I think I might have issue in my code, because I was previously doing re-authentication with same client, but now with multiple cloud support I will need to authenticate and obtain new provider/service client every time `Init()` is called from Velero. I noticed the GCP plugin is doing it this way.

I will try that and let you know when I finish.

**Comment by Lirt** [comment]
Hi again,

I fixed the issue by 2 steps:
1. I'm authenticating and getting new provider and service client on every `Init()` call.
2. I didn't update Velero CRDs correctly which caused issue in Velero to not see BSL as available.

Now I can see both BSL backups correctly and do backups for both BSLs almost correctly :smile:.

When I do backup command it reports that Status is Failed, but when I get/describe it, it is successful. Logs don't report any errors and I can see all objects in object storage:
```bash
$ velero backup create backup-ic-hrvart3-01 --namespace kube-system --storage-location caas-backup-ic-hrvart3 -w
Backup request "backup-ic-hrvart3-01" submitted successfully.
Waiting for backup to complete. You may safely press ctrl-c to stop waiting - your backup will continue in the background.
..............
Backup completed with status: Failed. You may check for more information using the commands `velero backup describe backup-ic-hrvart3-01` and `velero backup logs backup-ic-hrvart3-01`.
```

Could this be some client/CLI issue? 
```yaml
$ kubectl describe backup backup-ic-hrvart3-01

Name:         backup-ic-hrvart3-01
Namespace:    kube-system
Labels:       velero.io/storage-location=caas-backup-ic-hrvart3
Annotations:  velero.io/source-cluster-k8s-gitversion: v1.24.4
              velero.io/source-cluster-k8s-major-version: 1
              velero.io/source-cluster-k8s-minor-version: 24
API Version:  velero.io/v1
Kind:         Backup
Spec:
  Csi Snapshot Timeout:       10m0s
  Default Volumes To Restic:  false
  Hooks:
  Included Namespaces:
    *
  Metadata:
  Storage Location:  caas-backup-ic-hrvart3
  Ttl:               720h0m0s
Status:
  Completion Timestamp:  2022-11-11T09:12:38Z
  Expiration:            2022-12-11T09:12:22Z
  Format Version:        1.1.0
  Phase:                 Completed
  Progress:
    Items Backed Up:  1128
    Total Items:      1128
  Start Timestamp:    2022-11-11T09:12:22Z
  Version:            1
Events:               <none>
```

The functionality is correct, but if you have any idea if this CLI thing is something known, please let me know. If not I think we can close the issue as it doesn't affect overall functionality.

Thank you for your time and my apologies for opening "false" issue.

**Comment by blackpiglet** [comment]
Great!

**Comment by Lirt** [comment]
Hi, here is output of `velero debug` https://www.dropbox.com/s/idrx1b3ofrwghkc/bundle-2022-11-14-10-42-08.tar.gz?dl=0.

I tried to watch backups using `kubectl` and I can see that after backup is finished it is for a short time in status `Failed` and then converts to status `Completed`.

```yaml
Name:         backup-ic-hubudb3-05
Namespace:    kube-system
Labels:       velero.io/storage-location=caas-backup-ic-hubudb3
Annotations:  velero.io/source-cluster-k8s-gitversion=v1.24.4
              velero.io/source-cluster-k8s-major-version=1
              velero.io/source-cluster-k8s-minor-version=24

Phase:  Failed (run `velero backup logs backup-ic-hubudb3-05` for more information)

Errors:    0
Warnings:  0

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  caas-backup-ic-hubudb3

Velero-Native Snapshot PVs:  auto

TTL:  720h0m0s

Hooks:  <none>

Backup Format Version:  1.1.0

Started:    2022-11-14 10:50:21 +0100 CET
Completed:  2022-11-14 10:50:35 +0100 CET

Expiration:  2022-12-14 10:50:21 +0100 CET

Total items to be backed up:  1121
Items backed up:              1121

Velero-Native Snapshots: <none included>
```

Also when I execute `velero backup logs backup-ic-hubudb3-05 -n kube-system` - the last line (from STDERR) is `An error occurred: gzip: invalid checksum`.

If you need more debug logs or anything else, please let me know.

**Comment by Lirt** [comment]
Hi again, I will close the issue. The false backup state was solved by updating velero dependency in plugin code.

If I see issue reappearing I will open new issue, but you don't need to spend time here or keep tracking it.

Thank you very much for help.
